### PR TITLE
Display proper error message for WPP credit card failures

### PR DESCRIPTION
**Pull request by edra-bumci**

### CODE REVIEW

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 1]
```diff
@@ -420,7 +420,22 @@ function onError(error)
...
+    let cardFailureRequest =  jQuery.ajax({
+        url: ccControllerUrl,
+        data: {
+            action: "creditCardFailure",
+            "errors": errorList
+        }
+    });
+
+    cardFailureRequest
+        .done(()=> {
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Missing space after =>.](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 2]
```diff
@@ -114,4 +114,19 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+    /**
+     * Add proper error message on credit card failed payment
+     *
+     * @since 2.7.0
+     */
+    public function displayAjaxCreditCardFailure()
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [displayAjaxCreditCardFailure accesses the super-global variable $_SESSION.](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 3]
```diff
@@ -114,4 +114,19 @@ public function displayAjaxGetSeamlessConfig()
...
+    {
+        session_start();
+        $errorList = Tools::getValue('errors');
+        $notification = json_encode([
+        	'error' => $errorList
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 23
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 4]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+	/**
+	 * Add proper error message on credit card failed payment
+	 *
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 5]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
...
+	 * @since 2.7.0
+	 */
+	public function displayAjaxCreditCardFailure()
+	{
+		$errorList = Tools::getValue('errors');
+		$returnHtmlDOM = $this->context_service->redirectWithError($errorList, 'order');
+		$response = new Response($returnHtmlDOM,201);
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 6]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
...
+	/** @var ContextService */
+	private $context_service;
 
-        parent::initContent();
-    }
+	public function initContent()
+	{
+		$this->ajax = true;
+		$this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+		$this->context_service = new ContextService(\Context::getContext());
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 7]
```diff
@@ -420,7 +420,24 @@ function onError(error)
...
+        data: {
+            action: "creditCardFailure",
+            "errors": errorList
+        }
+    });
+
+    cardFailureRequest
+        .done(($response)=> {
+            document.open();
+            document.write($response);
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Unsafe call to document.write](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 8]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
...
-        $this->ajax = true;
-        $this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+	/** @var ContextService */
+	private $context_service;
 
-        parent::initContent();
-    }
+	public function initContent()
+	{
+		$this->ajax = true;
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 9]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
...
+
+	/**
+	 * Add proper error message on credit card failed payment
+	 *
+	 * @since 2.7.0
+	 */
+	public function displayAjaxCreditCardFailure()
+	{
+		$errorList = Tools::getValue('errors');
+		$returnHtmlDOM = $this->context_service->redirectWithError($errorList, 'order');
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 10]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
...
+		$response = new Response($returnHtmlDOM,201);
+		$response->send();
+	}
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 11]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+	/**
+	 * Add proper error message on credit card failed payment
+	 *
+	 * @since 2.7.0
+	 */
+	public function displayAjaxCreditCardFailure()
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 12]
```diff
@@ -420,7 +420,24 @@ function onError(error)
...
+    let cardFailureRequest =  jQuery.ajax({
+        url: ccControllerUrl,
+        data: {
+            action: "creditCardFailure",
+            "errors": errorList
+        }
+    });
+
+    cardFailureRequest
+        .done(($response)=> {
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Missing space before =>.](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 13]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
...
-    public function initContent()
-    {
-        $this->ajax = true;
-        $this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+	/** @var ContextService */
+	private $context_service;
 
-        parent::initContent();
-    }
+	public function initContent()
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 14]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+	/**
+	 * Add proper error message on credit card failed payment
+	 *
+	 * @since 2.7.0
+	 */
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 15]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
...
+	public function displayAjaxCreditCardFailure()
+	{
+		$errorList = Tools::getValue('errors');
+		$returnHtmlDOM = $this->context_service->redirectWithError($errorList, 'order');
+		$response = new Response($returnHtmlDOM,201);
+		$response->send();
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 16]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
...
-        $this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+	/** @var ContextService */
+	private $context_service;
 
-        parent::initContent();
-    }
+	public function initContent()
+	{
+		$this->ajax = true;
+		$this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 17]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
      */
     private $credit_card_vault_model;
 
-    public function initContent()
-    {
-        $this->ajax = true;
-        $this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+	/** @var ContextService */
+	private $context_service;
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 18]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
...
     }
+
+	/**
+	 * Add proper error message on credit card failed payment
+	 *
+	 * @since 2.7.0
+	 */
+	public function displayAjaxCreditCardFailure()
+	{
+		$errorList = Tools::getValue('errors');
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 19]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+	/**
+	 * Add proper error message on credit card failed payment
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 20]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
...
+	 *
+	 * @since 2.7.0
+	 */
+	public function displayAjaxCreditCardFailure()
+	{
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 21]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
      */
     private $credit_card_vault_model;
 
-    public function initContent()
-    {
-        $this->ajax = true;
-        $this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+	/** @var ContextService */
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 22]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+	/**
+	 * Add proper error message on credit card failed payment
+	 *
+	 * @since 2.7.0
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 23]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
...
+	private $context_service;
 
-        parent::initContent();
-    }
+	public function initContent()
+	{
+		$this->ajax = true;
+		$this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
+		$this->context_service = new ContextService(\Context::getContext());
+		parent::initContent();
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/helper/Service/ContextService.php [hunk 24]
```diff
@@ -57,7 +57,7 @@ public function setCart($cart)
     public function redirectWithError($errors, $controller_name)
     {
         $this->context->controller->errors = $errors;
-        $this->context->controller->redirectWithNotifications($this->context->link->getPageLink($controller_name));
+        return $this->context->controller->redirectWithNotifications($this->context->link->getPageLink($controller_name));
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/ContextService.php]: Codacy Issue found: [Line exceeds 120 characters; contains 122 characters](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 25]
```diff
@@ -114,4 +118,17 @@ public function displayAjaxGetSeamlessConfig()
 
         $response->send();
     }
+
+	/**
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php (hunk 5 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [No space found after comma in argument list](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/helper/Service/ContextService.php [hunk 26]
```diff
@@ -57,7 +57,9 @@ public function setCart($cart)
     public function redirectWithError($errors, $controller_name)
     {
         $this->context->controller->errors = $errors;
-        $this->context->controller->redirectWithNotifications($this->context->link->getPageLink($controller_name));
+        return $this->context->controller->redirectWithNotifications(
+        	$this->context->link->getPageLink($controller_name)
```
Comment (rinnhofer) [inline wirecardpaymentgateway/helper/Service/ContextService.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 27]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
      */
     private $credit_card_vault_model;
 
+    /** @var ContextService */
+    private $context_service;
+
     public function initContent()
-    {
+	{
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/creditcard.php [hunk 28]
```diff
@@ -26,13 +27,16 @@ class WirecardPaymentGatewayCreditCardModuleFrontController extends ModuleFrontC
...
         $this->ajax = true;
         $this->credit_card_vault_model = new CreditCardVault($this->context->customer->id);
-
+        $this->context_service = new ContextService(\Context::getContext());
         parent::initContent();
-    }
+	}
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/creditcard.php]: Codacy Issue found: [Spaces must be used to indent lines; tabs are not allowed](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 29]
```diff
@@ -420,7 +420,24 @@ function onError(error)
...
+        data: {
+            action: "creditCardFailure",
+            "errors": errorList
+        }
+    });
+
+    cardFailureRequest
+        .done(($response) => {
+            document.open();
+            document.write($response);
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Unsafe call to document.write](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/views/js/creditcard.js (hunk 29 above)
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Unsafe call to document.write](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 30]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
+     *
+     * @return boolean
+     * @throws Exception
+     * @since 2.7.0
+     */
+     private function determineErrorException($errorNotification)
+    {
+        if($errorNotification){
+            throw new Exception($errorNotification);
+    }
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Line indented incorrectly; expected 8 spaces, found 4](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 31]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
+        return true;
+    }
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Closing brace indented incorrectly; expected 5 spaces, found 4](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 32]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
+     *
+     * @param array $errorNotification
+     *
+     * @return boolean
+     * @throws Exception
+     * @since 2.7.0
+     */
+     private function determineErrorException($errorNotification)
+    {
+        if($errorNotification){
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Expected 1 space(s) after closing parenthesis; found 0](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 33]
```diff
@@ -418,9 +418,19 @@ function onError(error)
...
-    console.error("Form error:", error);
+    let $form = jQuery(Constants.PAYMENT_FORM_ID);
+    let $errorList = [];
+    error.errors.forEach(function(item){
+        $errorList.push(item.error.description);
+    });
+    let $input = jQuery("<input>").attr({
+        type: "hidden",
+        value: $errorList,
+        name: 'error-notification',
```
Comment (rinnhofer) [inline wirecardpaymentgateway/views/js/creditcard.js]: Codacy Issue found: [Strings must use doublequote.](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 34]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
+     * Check the error notification field to trigger an exception
+     *
+     * @param array $errorNotification
+     *
+     * @return boolean
+     * @throws Exception
+     * @since 2.7.0
+     */
+     private function determineErrorException($errorNotification)
+    {
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Opening brace indented incorrectly; expected 5 spaces, found 4](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 35]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
+     /**
+     * Check the error notification field to trigger an exception
+     *
+     * @param array $errorNotification
+     *
+     * @return boolean
+     * @throws Exception
+     * @since 2.7.0
+     */
+     private function determineErrorException($errorNotification)
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Line indented incorrectly; expected 4 spaces, found 5](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 36]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
+     *
+     * @param array $errorNotification
+     *
+     * @return boolean
+     * @throws Exception
+     * @since 2.7.0
+     */
+    private function determineErrorException($errorNotification)
+    {
+        if($errorNotification) {
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Expected 1 space(s) after closing parenthesis; found 0](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/controllers/front/payment.php [hunk 37]
```diff
@@ -57,6 +60,23 @@ public function postProcess()
...
```
Comment (rinnhofer) [inline wirecardpaymentgateway/controllers/front/payment.php]: Codacy Issue found: [Expected 1 space(s) after closing parenthesis; found 0](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

**Comment by rinnhofer** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- wirecardpaymentgateway/helper/AdditionalInformationBuilder.php  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/wirecard/prestashop-ee/pullRequest?prid=4985232&bid=16461462)

File: wirecardpaymentgateway/helper/AdditionalInformationBuilder.php [hunk 38]
```diff
@@ -176,12 +176,14 @@ public function createAccountHolder($cart, $type, $firstName = null, $lastName =
...
 
-        if (\Tools::strlen($address->phone)) {
-            $accountHolder->setPhone($address->phone);
+        if (\Tools::strlen(trim($address->phone))) {
+            $accountHolder->setPhone(trim($address->phone));
         }
 
-        if (\Tools::strlen($address->phone_mobile)) {
-            $accountHolder->setMobilePhone($address->phone_mobile);
+        if ($type === 'billing') {
```
Comment (ozada) [inline wirecardpaymentgateway/helper/AdditionalInformationBuilder.php]: Usage rough values increases technical doubt. Please replace it with constant.
You could also rewrite nested IFs  like:
if type=ADDRESS_TYPE_BILLING && trim(address.phoneMobile):
    setMobileNumber

File: wirecardpaymentgateway/models/Transaction.php [hunk 39]
```diff
@@ -765,6 +765,11 @@ public function isCaptureSettledTransitive()
         return $this->equals($this->getProcessedCaptureAmountTransitive(), $this->getAmount());
     }
 
+    private function isOrderStateCancel($orderState)
```
Comment (ozada) [inline wirecardpaymentgateway/models/Transaction.php]: Missing doc block
usage of (int)$x is 6x faster as intval($x)

File: wirecardpaymentgateway/models/Transaction.php [hunk 40]
```diff
@@ -789,6 +794,13 @@ public function updateOrder(
             $order->save();
             $updated = true;
         }
+
+        if (!$updated && $this->isOrderStateCancel($orderState)) {
```
Comment (ozada) [inline wirecardpaymentgateway/models/Transaction.php]: I think it's good opportunity now extract some logic to new functions. (ask @tomazpu or @flavius is that make value now)

File: wirecardpaymentgateway/views/js/creditcard.js [hunk 41]
```diff
@@ -418,9 +419,19 @@ function onError(error)
  * @param error
  * @since 2.4.0
  */
-function onFormError(error)
+function onSeamlessFormError(error)
 {
```
Comment (ozada) [inline wirecardpaymentgateway/views/js/creditcard.js]: It should be nice to check is error some kind of Array and not empty.

File: wirecardpaymentgateway/views/js/sepadirectdebit.js [hunk 42]
```diff
@@ -24,6 +26,8 @@ $(document).ready(function () {
     {
```
Comment (ozada) [inline wirecardpaymentgateway/views/js/sepadirectdebit.js]: #sepaCheck / #sepaConfirmButton / #payment-confirmation button -> could be extracted to constant

File: wirecardpaymentgateway/views/js/sepadirectdebit.js [hunk 43]
```diff
@@ -12,7 +12,9 @@ var sepaCheck = false;
 $(document).ready(function () {
     function processOrder()
     {
-        form.submit();
+        if ($("#sepaCheck").is(":checked")) {
+            form.submit();
+        }
     }
 
     function checkChange()
```
Comment (ozada) [inline wirecardpaymentgateway/views/js/sepadirectdebit.js]: missed doc block
Giving more reasonable name could be helpful.

File: wirecardpaymentgateway/views/js/sepadirectdebit.js [hunk 44]
```diff
@@ -12,7 +12,9 @@ var sepaCheck = false;
 $(document).ready(function () {
     function processOrder()
     {
-        form.submit();
+        if ($("#sepaCheck").is(":checked")) {
```
Comment (ozada) [inline wirecardpaymentgateway/views/js/sepadirectdebit.js]: #sepaCheck define constant

File: wirecardpaymentgateway/views/js/creditcard.js (hunk 41 above)
Comment (flavius) [reply wirecardpaymentgateway/views/js/creditcard.js]: Looks good.

File: wirecardpaymentgateway/helper/AdditionalInformationBuilder.php (hunk 38 above)
Comment (edra-bumci) [reply wirecardpaymentgateway/helper/AdditionalInformationBuilder.php]: @ozada outdated

File: wirecardpaymentgateway/models/Transaction.php (hunk 39 above)
Comment (edra-bumci) [reply wirecardpaymentgateway/models/Transaction.php]: @ozada outdated

File: wirecardpaymentgateway/models/Transaction.php (hunk 40 above)
Comment (edra-bumci) [reply wirecardpaymentgateway/models/Transaction.php]: @ozada outdated

File: wirecardpaymentgateway/views/js/sepadirectdebit.js (hunk 43 above)
Comment (edra-bumci) [reply wirecardpaymentgateway/views/js/sepadirectdebit.js]: @ozada outdated

File: wirecardpaymentgateway/views/js/sepadirectdebit.js (hunk 44 above)
Comment (edra-bumci) [reply wirecardpaymentgateway/views/js/sepadirectdebit.js]: @ozada outdated

**Pull request merged**